sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgBuilder.cs

```diff
@@ -0,0 +1,126 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2018 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.Protobuf.Ucfg;
+
+namespace SonarAnalyzer.ControlFlowGraph.CSharp
+{
+    public class UcfgBuilder
+    {
+        private readonly UcfgObjectFactory objectFactory;
+        private readonly UcfgBlockIdProvider blockIdProvider;
+        private readonly UcfgInstructionFactory instructionFactory;
+
+        public UcfgBuilder(SemanticModel semanticModel)
+        {
+            objectFactory = new UcfgObjectFactory();
+            blockIdProvider = new UcfgBlockIdProvider();
+            instructionFactory = new UcfgInstructionFactory(semanticModel, objectFactory);
+        }
+
+        public UCFG Build(SyntaxNode syntaxNode, IMethodSymbol methodSymbol, IControlFlowGraph cfg)
+        {
+            var ucfg = objectFactory.CreateUcfg(syntaxNode, UcfgMethod.Create(methodSymbol));
+
+            ucfg.BasicBlocks.AddRange(cfg.Blocks.Select(CreateBasicBlock));
+
+            ucfg.Parameters.AddRange(methodSymbol.GetParameters().Select(p => p.Name));
+
+            if (syntaxNode is BaseMethodDeclarationSyntax methodDeclaration &&
+                TaintAnalysisEntryPointDetector.IsEntryPoint(methodSymbol))
+            {
+                var entryPointBlock = CreateEntryPointBlock(methodDeclaration, methodSymbol, blockIdProvider.Get(cfg.EntryBlock));
+                ucfg.BasicBlocks.Add(entryPointBlock);
+                ucfg.Entries.Add(entryPointBlock.Id);
+            }
+            else
+            {
+                ucfg.Entries.Add(blockIdProvider.Get(cfg.EntryBlock));
+            }
+
+            return ucfg;
+        }
+
+        private BasicBlock CreateBasicBlock(Block block)
```

Comment:
I thought we said on Friday that we will have`UcfgBuilder`, `UcfgBlockBuilder`, `UcfgInstructionBuilder`?

Reply:
I'll leave splitting UcfgBuilder to UcfgBlockBuilder and UcfgBuilder to you and @duncanp-sonar :)

Reply:
In general the *Builder classes imply implementing Builder pattern, which we are not. I could change the UcfgBuilder to CfgToUcfgConverter but the name is clumsy...

Reply:
I started to disagree when saw that having two classes sharing the same state and having two methods each will not improve a lot. It is not a lot of work to separate, I just think the current class is not that bad to require changes.